### PR TITLE
[privacy] implement octojoin

### PR DESCRIPTION
**Pull request by 1440000bytes**
This pull request implements a proof of concept for octojoin. I have previously tried this privacy feature with bitcoin core and [cake wallet](https://github.com/cake-tech/cake_wallet/pull/3167).

Read more: [octojoin.github.io](https://octojoin.github.io)

### CODE REVIEW

File: screen/send/SendDetails.tsx [hunk 1]
```diff
@@ -504,11 +508,74 @@ const SendDetails = () => {
     [setParams, wallet],
   );
 
+  const parseOctojoinDestinations = (): { addressList: string[]; isSilentPayment: boolean; amountSats: number } => {
+    const first = addresses[0];
```
Comment (GladosBlueWallet) [inline screen/send/SendDetails.tsx:516]: This threshold is below OCTOJOIN_DUST_THRESHOLD, so amounts between 501 and 546 sats are accepted here but decompose into zero payment outputs later. Worse, non-multiples above that still silently donate the remainder into fees. The validator and the decomposer need to agree before this starts eating coins.

**Review by GladosBlueWallet — COMMENTED** [review]
Octojoin is the kind of idea that sounds harmless right up until the input validator and the amount decomposer disagree about what counts as a valid payment. Your 501-546 sat cases slip through the gate and then collapse into no outputs or dust-eating fee behavior. Remarkably efficient at creating user confusion.

**Comment by 1440000bytes** [comment]
@ncoelho Can you please review this pull request and share some feedback?

**Comment by ncoelho** [comment]
> @ncoelho Can you please review this pull request and share some feedback? 

Add some screenshots of the user flows and explain what you are trying to do please.

File: class/octojoin.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  const selectedSwapped = swappedUtxos.slice(0, requiredSwapped);
+  const selectedOther: T[] = [];
+
+  const sortedOther = [...otherUtxos].sort((a, b) => b.value - a.value);
+
+  let currentTotalValue = selectedSwapped.reduce((sum, u) => sum + u.value, 0);
+
+  for (const other of sortedOther) {
```
Comment (GladosBlueWallet) [inline class/octojoin.ts]: **[CRITICAL]** Your loop hoards multiple non-octojoin UTXOs. Octojoin wants exactly one sender coin plus swapped decoys — congratulations, you just built a fingerprinting machine with extra steps.

File: class/octojoin.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  if (currentTotalValue < targetAmount) {
+    throw new Error(
+      `Insufficient funds. Total selected: ${(currentTotalValue / 100000000).toFixed(8)} BTC, ` +
+        `Target: ${(targetAmount / 100000000).toFixed(8)} BTC.`,
+    );
+  }
+
+  return {
```
Comment (GladosBlueWallet) [inline class/octojoin.ts:130]: **[HIGH]** `selection.all` has no ceiling at `numInputs`. User sets 3, you spend 4+. Wrong tx shape, wrong anonymity set — the test subject lied on the intake form.

File: class/octojoin.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  let remaining = amountInSatoshis;
+
+  for (const denom of OCTOJOIN_STANDARD_DENOMINATIONS) {
+    while (remaining >= denom) {
+      denominations.push(denom);
+      remaining -= denom;
+    }
+  }
+
+  if (remaining > OCTOJOIN_DUST_THRESHOLD) {
```
Comment (GladosBlueWallet) [inline class/octojoin.ts:25]: **[HIGH]** Dust remainder is discarded from denominations with no alert. Enter 100500 sats, recipient gets 100000. Silent underpayment — my favorite kind of surprise, for the recipient.

File: screen/send/SendDetails.tsx [hunk 5]
```diff
@@ -504,11 +508,74 @@ const SendDetails = () => {
...
+      return loc.send.octojoin_outputs;
+    }
+    const { addressList, isSilentPayment, amountSats } = parseOctojoinDestinations();
+    if (!amountSats || amountSats <= 500) {
+      return loc.send.details_amount_field_is_not_valid;
+    }
+    if (!feeRate || parseFloat(feeRate) < 0) {
+      return loc.send.details_fee_field_is_not_valid;
+    }
+    if (isSilentPayment) {
```
Comment (GladosBlueWallet) [inline screen/send/SendDetails.tsx:547]: **[HIGH]** Silent-payment branch skips `numOutputs` validation entirely. UI knob is decorative; `planOctojoin` emits one output per denomination regardless. Cake had a protocol. You have a toggle.

File: screen/send/SendDetails.tsx [hunk 6]
```diff
@@ -504,11 +508,74 @@ const SendDetails = () => {
...
+    const numInputs = parseInt(octojoinNumInputs, 10);
+    if (!numInputs || numInputs < OCTOJOIN_MIN_INPUTS) {
+      return loc.send.octojoin_inputs;
+    }
+    const numOutputs = parseInt(octojoinNumOutputs, 10);
+    if (!numOutputs || numOutputs < OCTOJOIN_MIN_OUTPUTS) {
+      return loc.send.octojoin_outputs;
+    }
+    const { addressList, isSilentPayment, amountSats } = parseOctojoinDestinations();
+    if (!amountSats || amountSats <= 500) {
```
Comment (GladosBlueWallet) [inline screen/send/SendDetails.tsx]: **[HIGH]** `validateOctojoin` never checks `decomposeAmount(amountSats)` sums to `amountSats`. Dust truncation passes validation, then underpays on broadcast. Science requires consent.

File: tests/unit/octojoin.test.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+  it('isOctojoinMemo matches case-insensitively and within longer notes', () => {
+    assert.strictEqual(isOctojoinMemo('Octojoin 1'), true);
+    assert.strictEqual(isOctojoinMemo('octojoin 2'), true);
+    assert.strictEqual(isOctojoinMemo('my OCTOJOIN swap'), true);
+    assert.strictEqual(isOctojoinMemo('Normal TX'), false);
+    assert.strictEqual(isOctojoinMemo(''), false);
+    assert.strictEqual(isOctojoinMemo(undefined), false);
+  });
+
+  it('selectOctojoinUtxos isolates octojoin-tagged coins from the rest', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/octojoin.test.ts:66]: **[HIGH]** No test that selection uses exactly one non-octojoin input or rejects excess inputs. The critical multi-coin bug could waltz past CI forever. Testing theater.

File: tests/unit/octojoin.test.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      plan.paymentTargets.reduce((s, t) => s + t.value, 0),
+      300000,
+    );
+    assert.strictEqual(plan.totalInput, 1000000);
+    // payment is decomposed across the addresses round-robin: 0.002 then 0.001
+    assert.strictEqual(plan.paymentTargets[0].value, 200000);
+    assert.strictEqual(plan.paymentTargets[1].value, 100000);
+  });
+
+  it('expands a silent payment address into one output per denomination', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/octojoin.test.ts:222]: **[HIGH]** Silent-payment path only asserts `planOctojoin` targets — never builds/signs an SP octojoin tx. Primary use case untested; mocks need not apply when you skip the hard part entirely.

File: class/octojoin.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  const requiredSwapped = numInputs - 1;
+
+  if (swappedUtxos.length < requiredSwapped) {
+    throw new Error(`Not enough 'octojoin' coins. You need at least ${requiredSwapped}, but only found ${swappedUtxos.length}.`);
+  }
+  if (otherUtxos.length === 0) {
+    throw new Error('Requires at least 1 non-octojoin coin.');
+  }
+
+  const selectedSwapped = swappedUtxos.slice(0, requiredSwapped);
```
Comment (GladosBlueWallet) [inline class/octojoin.ts]: **[MEDIUM]** `swappedUtxos.slice(0, requiredSwapped)` grabs wallet order, not best fit. Wrong pair plus one insufficient coin triggers the multi-other bug. Selection roulette.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**PoC privacy feature** (octojoin: swapped UTXO inputs + denomination-split outputs + silent payments). Core idea is sound; the **input-selection chamber** violates the protocol's one-sender-coin rule, can inflate input count past your dial, and silently shaves sats off payments. Tests exercise happy paths but skip the edge cases where this experiment actually explodes. Commits and PR hygiene also failed CONTRIBUTING checks.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:3` — **[MEDIUM]** Four commits say `octojoin: add …`. CONTRIBUTING demands `ADD:`/`TST:` prefixes. Lab protocol violation — I do appreciate the irony.
- `CONTRIBUTING.md:33` — **[MEDIUM]** PR body has no screenshot or video of the Send UI. CONTRIBUTING requires visual proof. I cannot admire your octojoin panel through telepathy. Yet.

**Comment by 1440000bytes** [comment]
> > @ncoelho Can you please review this pull request and share some feedback?
> 
> Add some screenshots of the user flows and explain what you are trying to do please.

I was unable to use bluewallet with testnet last time. I will try again this weekend and share the screenshots.

**Comment by 1440000bytes** [comment]
<details>

<summary>Screenshots</summary>







</details>

<details>

<summary>Demo Video</summary>

https://github.com/user-attachments/assets/a9c06403-4044-406c-9cd5-2967ce93137f

</details>
